File: LeetCode/1_two_sum.py
""" February 2019
runtime: 36ms [99th percentile]
memory: 7.9Mb [great!]
"""


class Solution:
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        hash_table = dict()
        for ind, elem in enumerate(nums):
            result = target-nums[ind]
            if result in hash_table:
                return [hash_table[result], ind]
            hash_table[elem]=ind

File: LeetCode/1_two_sum.cpp
/* February 2019
runtime: 4ms [99th]
memory: 1.6Mb [the best?]
*/

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> res = {0};
        unordered_map<int,int> hash_map;
        int remainder=0;
        for(int index=0; index < nums.size(); index++){
            remainder = target - nums[index];
            if(hash_map.find(remainder) == hash_map.end())
                hash_map[nums[index]] = index;
            else{
                res = {hash_map[remainder], index};
            }
        }
        return res;
    }
};